### DISCUSSION TITLE
Future development

### DISCUSSION THREAD

**Discussion by guanzhangrtk** [yume-chan/ya-webadb#341] [General]
Hi @yume-chan I realize that this a side project for you, but curious if you can share some thoughts on its future development. Do you have any features you have planned that you want to implement when you get the time? If you list them here perhaps other people who are interested can help out as well.

Thank you!

**Comment by yume-chan** [comment]
Currently my TODO list contains only one thing:

- [ ] support more message types in the scrcpy compatible client (mouse scroll, etc)

I'm not an Android developer or Android hacker, I even didn't know what ADB or Android can do before starting this project, so I have no idea what features are still lacking.

**Reply by guanzhangrtk** [reply]
I guess off the top of my head is, can we improve tinyh264 performance such that it doesn't use as much CPU -- is it possible to use hardware acceleration?

The other thing of interest is can we create some macro/scripts/keymapper for use with Android devices. There are some automation tools available for Android devices but most of them are not Open Source. For automation/testing, there are some Open Source projects like Airtest and Appium however they are more geared towards testers than regular end-users.

Thanks!

**Reply by yume-chan** [reply]
> is it possible to use hardware acceleration?

There had been multiple attempts (by me, by others) to use MSE ([Media Source Extensions](https://developers.google.com/web/fundamentals/media/mse/basics)) to enable hardware accelerated video decoding, all failed. Not clear why, but browsers just can't playback the remuxed video file in consistent speed.

> The other thing of interest is can we create some macro/scripts/keymapper for use with Android devices.

Can you describe the feature in more detail?

**Reply by guanzhangrtk** [reply]
> > The other thing of interest is can we create some macro/scripts/keymapper for use with Android devices.
> 
> Can you describe the feature in more detail?

Sure! Take a look at this project: https://github.com/barry-ran/QtScrcpy#customized-key-mapping

Basically it allows one to use keyboard and mouse to interact with the Android device using `scrcpy`. This is the "keymapping" part.

A macro would allow you to string a few operations in sequence. For example there are 3 buttons on screen at coordinates X, Y and Z, so an example macro could be to tap the 3 buttons in sequence and mapping it to the keyboard key "A". This would allow automated gameplay using these features.

I'm not sure whether these are possible with `node.js`.

BTW in case you're interested, I run a Discord server on Android Emulators where we discuss topics on Android and other related tools (your tool is listed in our resources). This way we can discuss more interactively and then we can summarize our discussions here afterwards: https://discord.gg/mRpT4Qq

Thanks!

**Reply by mattwakeman** [reply]
This python script shows how to make 'long' screenshots using ADB to take a screenshot, then send input messages to scroll the screen before taking another. It then uses pillow to stitch them in together into one image

https://github.com/google/android-wear-stitch-script
